### ISSUE TITLE
Upload file from stdin

### ISSUE THREAD

**Issue by nhoffman** [timvisee/ffsend#39]
Hello, I'm wondering if you have considered support for reading from stdin or a named pipe to avoid having to write a file to disk. For example, all of the following fail with various errors:

```
echo foo | ffsend upload --name foo.txt
echo foo | ffsend upload --name foo.txt -
ffsend upload --name foo.txt <(echo foo)
```

Thanks!
Noah

**Comment by timvisee** [comment]
Yes, and I would like to implement it.

The problem is however, that `ffsend` must know the exact size of the file to upload which is a limitation caused by Firefox Send. So, there are two options I believe.
The first being to add a CLI parameter the user can set the exact file size with. I'd argue that this is seldom useful, and it's quite error prone because if the file size is incorrect the upload will be malformed.
The second solution is to write all data from `stdin` to a temporary file first, then upload as normal when the file size is known. I would like to implement this using the latter method some time, but I haven't had the time to do so yet.

**Comment by nhoffman** [comment]
Thanks a lot for the rapid response (and for this tool!). I agree that it wouldn't be much use to ask the user to provide the file size at the time of upload. If I find myself dying for the feature, it would be simple enough to write a python wrapper that provides this functionality in the interim.

**Comment by timvisee** [comment]
Yes, that's indeed perfectly doable. If you're doing so, it would be awesome if you could share the script for it.

I might take a look at implementing this (for uploading) this weekend. Although I'm not quite sure if have time for it yet.

**Comment by ignatenkobrain** [comment]
Can't you just read whole stdin and check size of it before upload? I might be able to supply patch. Stay tuned.

**Comment by timvisee** [comment]
@ignatenkobrain Yes, that is probably the proper approach.

Writing it to a temporary file first is probably much easier, as you can pass that temporary file to the `upload` command as if you're uploading a file. Then, everything should be handled automatically.

I like the idea of using `-` as indicator to use from `stdin`. And it might be a good idea to require `--name`, so users have to set the name of the uploaded file explicitly.

**Comment by ignatenkobrain** [comment]
Requiring more things sounds bad. In Fedora we are exploring possibilities for replacing our hosted pastebin and ffsend is one of ideas to consider... So if it asks for more than just providing text is just bad UX.

**Comment by ignatenkobrain** [comment]
Writing is problematic since you might be in RO environment. And even tmpfs sometimes might not exist. After all you just want to upload something.

**Comment by timvisee** [comment]
@ignatenkobrain 

> Requiring more things sounds bad. In Fedora we are exploring possibilities for replacing our hosted pastebin and ffsend is one of ideas to consider... So if it asks for more than just providing text is just bad UX.

Fair enough. A default name would be reasonable.

> Writing is problematic since you might be in RO environment. And even tmpfs sometimes might not exist. After all you just want to upload something.

You're right. I'm using limited environments myself where writing is _not_ possible. I'm wondering how adaptable the current upload implementation is for this, though.

**Comment by ignatenkobrain** [comment]
@timvisee so I took quick look at ffsend-api, so here are my thoughts:

We have Reader struct and we can pass size and filename around. So if we pass stdin as an Reader, it should work. Although we still need to know size of stdin, which can be done for some OS, but not others. I'm not sure how ECE is working and whether it is possible to just continue encrypting until you encounter end of stream.

So to sum it up, I think on linux and BSD (at least) you can rely on size which you can get from fs::metadata, but for other OS just write a temporary file and upload it...
We need to come up with some interface where you pass

**Comment by timvisee** [comment]
@ignatenkobrain 

Let me explain a little about the [`pipe`](https://github.com/timvisee/ffsend-api/tree/master/src/pipe) system I've built, as it might help you understand implementation possibilities better. It might be a little difficult to understand, just glancing at the code. I wanted to keep it as dynamic and adaptable as possible. The latest crypto change basically made my previous implementation unusable, while this new implementation should be adaptable to future changes.

The basic idea is that you can attach a few `Pipe`s together, each `Pipe` having a specific task, with a reader/writer on both ends of the chain. So, the ECE crypto used for the latest Firefox Send version is a `Pipe`. Which you could attach to an optional archiving type when the `-a` flag is given. You then create a reader/writer for this chain of pipes, depending on what is desired. When uploading, a `File` (being a `Reader`) is then streamed through this pipe, and is written to the remote host. I hope this is somewhat understandable.

Now, please note that archiving being a `Pipe` was just an example in this case to help explain the above. Because the size of the final archive must be known, archiving is currently done before creating these pipes for encryption to upload, after which the archive is passed as a regular file.

Effectively, everything that is a `Reader` or a `Writer` should be usable for uploading/downloading, with some minor limitations such as the actual upload size that must be known.

---

> I'm not sure how ECE is working and whether it is possible to just continue encrypting until you encounter end of stream.

No, sadly the file size has to be known for ECE encryption in this case.

> I think on linux and BSD (at least) you can rely on size which you can get from fs::metadata

That's sounds rather cool. Yes, using that length value would be brilliant. Are you sure this is indeed the case though? [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html) always returns a value for the size. What would happen if the size is not known, would obtaining the [`Metadata`](https://doc.rust-lang.org/std/fs/struct.Metadata.html) instance fail before hand?

> but for other OS just write a temporary file and upload it

Once there's an easy method for attaching `stdin` to the upload pipeline, it shouldn't be too hard to attach any buffer I'm thinking. In that case, collecting `stdin` into an internal [buffer](https://docs.rs/bytes/0.4.12/bytes/struct.BytesMut.html) until `EOF` is reached, would be enough. Then, the input size is known and the buffer could be given to the upload pipeline as a reader.

Hope this makes sense. I hope I'm not making it sound more complicated than it is.

**Comment by tjoanp** [comment]
> The problem is however, that `ffsend` must know the exact size of the file to upload, which is a limitation introduced by Firefox Send

Considering Firefox Send is no more, would it be possible to change the requirements on the server side now?


Btw, thanks for an awesome software!

Best regards

**Comment by timvisee** [comment]
Though it's definitely possible, it would require significant changes in the way cryptography is done. This would also introduce a breaking change.

Knowing the file size is a requirement for the current crypto implementation.

You're saying that you'd like to use it as pastebin. Collecting all of stdin before uploading to determine its size would definitely work here. It's a solution I've suggested before, but this has not been implemented yet.

Maybe this is something I can implement this week. My planning is quite volatile though, so I'm not sure. Feel free to give it a try yourself!

**Comment by timvisee** [comment]
I've implemented this as part of `ffsend` `v0.2.71`. Provide `-` as file name in order to read from `stdin`.

```bash
echo "123" | ffsend upload -
echo "123" | ffsend upload --name=file.txt -
```

The current implementation isn't super efficient though. It stores the full contents in memory, in a temporary file, and doesn't have any limits. So don't try to pipe `/dev/random` into it. This is something I'll improve upon later. But, it definitely works.

This update should be available as soon as this release pipeline succeeds: https://gitlab.com/timvisee/ffsend/-/pipelines/284016357
